### ISSUE TITLE
Refactor optimize step

### ISSUE THREAD

**Issue by mlahariya** [pasqal-io/perceptrain#18]
### Describe the feature

Issue coming from qadence. 
https://github.com/pasqal-io/qadence/issues/732
https://github.com/pasqal-io/qadence/issues/631

### Would you like to work on this issue?

None

**Comment by mlahariya** [comment]
@debrevitatevitae @DanieleCucurachi - Lets discuss the optimize step in this issue.

**Comment by debrevitatevitae** [comment]
## Requirements
1. Optimize step should support sampling - for which dataloaders should be passed. (eg evolutionary sampling)
2. It should support both gradient based and gradient free optimizers 
3. It should support different loss functions - also different loss balancing techniques
